### PR TITLE
feat: add Avro Schema input processor

### LINKED ISSUE #1741: Add Avro support
#### Reason/Context
It would be great to have support for the Avro standard as supported inputs: https://avro.apache.org/

To support it we need to add a new input processor, which you can read about how to add here: https://github.com/asyncapi/modelina/blob/master/docs/contributing.md#adding-a-new-input-processor

The TLDR; is that we want to convert [their structure](https://avro.apache.org/docs/1.11.1/specification/) into our [Meta Models](https://modelina.org/docs/internal-model). Feel free to document the limitations of the implementation as needed in an input document: https://github.com/asyncapi/modelina/tree/master/docs/inputs

You can use the following Avro input document for testing (or come up with your own):
```json
{
    "name": "Person",
    "namespace": "com.company",
    "type": "record",
    "fields": [
      {"name": "name", "type": "string", "example": "Donkey", "minLength": 0},
      {"name": "serialNo", "type": "string", "minLength": 0, "maxLength": 50},
      {"name": "email", "type": ["null","string"], "example": "donkey@asyncapi.com", "pattern": "^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$"},
      {"name": "age", "type": ["null", "int"], "default": null, "example": 123, "exclusiveMinimum": 0, "exclusiveMaximum": 200},
      {
        "name": "favoriteProgrammingLanguage",
        "type": {"name": "ProgrammingLanguage", "type": "enum", "symbols": ["JS", "Java", "Go", "Rust", "C"], "default": "JS"}
      },
      {"name": "certifications", "type": {"type": "array", "items": "string", "minItems": 1, "maxItems": 500, "uniqueItems": true}},
      {
        "name": "address",
        "type": {
          "name": "Address",
          "type": "record",
          "fields": [
            {"name": "zipcode", "type": "int", "example": 53003},
            {"name": "country", "type": ["null", "string"]}
          ]
        }
      },
      {"name": "weight", "type": "float", "example": 65.1, "minimum": 0, "maximum": 500},
      {"name": "height", "type": "double", "example": 1.85, "minimum": 0, "maximum": 3.0},
      {"name": "someid", "type": "string", "logicalType": "uuid"}
    ]
}
```

Happy to help further if you get stuck on something :v:

### PR DESCRIPTION
**Pull request by akkshitgupta**
## Description
Add Avro Schema support.

## Related Issue 
fixes #1741 

## Checklist
- [x] The code follows the project's coding standards and is properly linted (`npm run lint`).
- [x] Tests have been added or updated to cover the changes.
- [x] All tests pass successfully locally.(`npm run test`).
- [x] Documentation has been updated to reflect the changes.

### CODE REVIEW

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [inline src/processors/AvroSchemaInputProcessor.ts:34]: This function should inspect the input for patterns that form an Avro schema, i.e. figure out if any properties are required in the root object and check for those. If none you can use the optional properties and say "if x amount of properties are found, its avro"

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [inline src/processors/AvroSchemaInputProcessor.ts:50]: Not sure if you need to reflect names 🤔 Only necessary if the user does does define all the necessary namings through the standard. I.e. you want to generate an enum, but the avro schema does not have a name for the enum.

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [inline src/processors/AvroSchemaInputProcessor.ts:56]: You dont need to convert Avro to common model only convert it to meta model directly 🙂 

This means that you build the meta model from the Avro schema: https://modelina.org/examples?selectedExample=meta-model

**Review by jonaslagoni — CHANGES_REQUESTED** [review]

File: src/processors/AvroSchemaInputProcessor.ts
Comment (akkshitgupta) [reply src/processors/AvroSchemaInputProcessor.ts:34]: As per the [documentation](https://avro.apache.org/docs/1.11.1/specification/#schema-declaration), Avro Schema has a `type` property at the root level, for which I implemented line 30. Moreover, Avro Schema can be defined as a `JSON string`, `JSON Object` or `JSON Array`. Covering all the cases in here.

File: src/processors/AvroSchemaInputProcessor.ts
Comment (akkshitgupta) [reply src/processors/AvroSchemaInputProcessor.ts:50]: `reflectSchemaNames` method is adopted to handle the `fields` property in case of `type: record` as follows:

```
{
  "type": "record",
  "name": "Example",
  "doc": "A simple name (attribute) and no namespace attribute: use the null namespace (\"\"); the fullname is 'Example'.",
  "fields": [
    {
      "name": "inheritNull",
      "type": {
        "type": "enum",
        "name": "Simple",
        "doc": "A simple name (attribute) and no namespace attribute: inherit the null namespace of the enclosing type 'Example'. The fullname is 'Simple'.",
        "symbols": ["a", "b"]
      }
    }, {
      "name": "explicitNamespace",
      "type": {
        "type": "fixed",
        "name": "Simple",
        "namespace": "explicit",
        "doc": "A simple name (attribute) and a namespace (attribute); the fullname is 'explicit.Simple' (this is a different type than of the 'inheritNull' field).",
        "size": 12
      }
    }, {
      "name": "fullName",
      "type": {
        "type": "record",
        "name": "a.full.Name",
        "namespace": "ignored",
        "doc": "A name attribute with a fullname, so the namespace attribute is ignored. The fullname is 'a.full.Name', and the namespace is 'a.full'.",
        "fields": [
          {
            "name": "inheritNamespace",
            "type": {
              "type": "enum",
              "name": "Understanding",
              "doc": "A simple name (attribute) and no namespace attribute: inherit the namespace of the enclosing type 'a.full.Name'. The fullname is 'a.full.Understanding'.",
              "symbols": ["d", "e"]
            }
          }
        ]
      }
    }
  ]
}
```

Implemented by observing the `anyOf`, `oneOf` and `allOf` cases of `JSON-Schema`. Couldn't figure out the exact nature of `reflectSchemaName` method.

Comment (jonaslagoni) [reply src/processors/AvroSchemaInputProcessor.ts:50]: You always have `name` so dont think its necessary to reflect anything 🙂

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [reply src/processors/AvroSchemaInputProcessor.ts:28]: It needs to be more precise because the JSON Schema also has `type` and how it can be defined. 

I would add the following:
- Check for `name` (it might not be there)
- Check the `type` property for null, boolean, int, long, float, double, bytes, string, records, enums, arrays, maps, unions and fixed

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [inline src/processors/AvroSchemaInputProcessor.ts:39]: You cannot use this function at all here as it's specific for the use of `CommonModel` types. You have to build your own function that iterates over the input building all relevant Meta Models that the input represents.

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Left some more comments

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [reply src/processors/AvroSchemaInputProcessor.ts:39]: For example when encountering `record` type, generate a `ObjectMetaModel`, `enum` generate a `EnumMetaModel` 🙂

**Comment by akkshitgupta** [comment]
since it is required to obtain the meta model directly from the Avro Schema bypassing the common model. I am trying to figure out how a schema is using common model in between to obtain the meta model. I read the different [meta models](https://modelina.org/docs/internal-model) but stuck at the implementation of it. 

Screenshot

**Comment by jonaslagoni** [comment]
What are you currently stuck on?

**Comment by akkshitgupta** [comment]
Getting started with it. I was reading the code base and connecting some of the dots for reverse engineering to find the way backward from meta model. but unfortunately could not use the meta models. perhaps failing to understand the flow of code.

File: src/processors/AvroSchemaInputProcessor.ts
Comment (akkshitgupta) [reply src/processors/AvroSchemaInputProcessor.ts]: Can you please guide me to test the generated data model from Avro Schema. I tried the same using playground UI in local development but could not obtain the appropriate result since it only works for JSON Schema or AsyncAPI, not sure about it.

**Comment by jonaslagoni** [comment]
@akkshitgupta looks like you got figured out some of it, let me know if you have anything blocking you 😄

**Comment by akkshitgupta** [comment]
Hey @jonaslagoni I forgot to ping you in this particular [comment](https://github.com/asyncapi/modelina/pull/1753#discussion_r1485133926). can you please go through this once. I am trying to test the processor but stuck at the bindings.
image

Also wanna confirm that whether we are going with the same `.json` file extension or gonna adopt the `.avsc` file extension and add a support for that as well. :)

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [inline src/helpers/AvroToMetaModel.ts:19]: I would definitely add some tests for this file to make it easier to assess how it behaves :v:

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [inline src/helpers/AvroToMetaModel.ts:36]: Dont think Avro has discriminator 🤔

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [inline src/helpers/AvroToMetaModel.ts:39]: Dont think avro has format either.

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [inline src/helpers/AvroToMetaModel.ts:211]: Seems like you are reusing incorrect implementation here from JSON Schema input 🤔

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Overall it looks like you are on the right track! Just make sure that you are not reusing code that is ONLY relevant for JSON Schema inputs and focus on what is possible with Avro and do so with tests :v:

If I where you I would take the example in the issue and add it as you star to guide towards, if the MetaModel that is being created does not match what you expect it means there is a bug somewhere :v: Also I would break down the example and add individual tests for each small inputs where possible to create better tests 😄

**Comment by jonaslagoni** [comment]
What those `bindings` are I have no idea 🤨 Are you using node v18?

File: src/helpers/AvroToMetaModel.ts
Comment (akkshitgupta) [reply src/helpers/AvroToMetaModel.ts:39]: May be I misunderstood the `ModelOptions`. From the code, I concluded that these `keys` like `discriminator`, `format`, `const`, etc. are required to get the final Meta Model and not related to the `JSON Schema` format.
I will make the necessary changes as I got my mistake.

Comment (jonaslagoni) [reply src/helpers/AvroToMetaModel.ts:39]: Yea, they are definitely only optional 👍

File: src/helpers/AvroToMetaModel.ts
Comment (akkshitgupta) [reply src/helpers/AvroToMetaModel.ts:19]: Tests are in progress and would push that by today.

**Comment by jonaslagoni** [comment]
Hey @akkshitgupta, have you had a chance to take a look at the review comments?

**Comment by akkshitgupta** [comment]
Apologies for the delay @jonaslagoni I am working on it. some tests broke yesterday figuring out what happened.

**Comment by akkshitgupta** [comment]
Hey @jonaslagoni, I am stuck at the process of converting an array as I could not figure out the `keyword` which would represent the entire array.

image

here the default keyword is used to represent an empty array.

**Comment by jonaslagoni** [comment]
@akkshitgupta in AsyncAPI spec-json-schema repo we have this JSON Schema document that can validates Avro documents: https://github.com/asyncapi/spec-json-schemas/blob/master/common/avroSchema_v1.json

I suggest we focus on that, because yea, the spec is not clear on what `items: the schema of the array’s items.` can contain.

According to the JSON Schema: https://github.com/asyncapi/spec-json-schemas/blob/35ea79055d09bdd319ed816e4ec169a12f0bb7a5/common/avroSchema_v1.json#L225 it can contain `types`, which is: https://github.com/asyncapi/spec-json-schemas/blob/35ea79055d09bdd319ed816e4ec169a12f0bb7a5/common/avroSchema_v1.json#L14

**Comment by akkshitgupta** [comment]
I have some doubts regarding [`spec-json-schema`](https://github.com/asyncapi/spec-json-schemas/blob/master/common/avroSchema_v1.json)
- are we gonna implement `spec-json-schema` validation instead of [Avro Schema](https://github.com/asyncapi/modelina/blob/73f47d7f7a326cb7f40e71658574c3b9e9510f71/src/models/AvroSchema.ts) defined here or we just need to implement the scenario/concept used there for `avro array` and use that implementation while getting the Array Meta Model
- not understood this particular thing, what you are referring to in here
    > I suggest we focus on that

**Note**: I have some academic exams and I might be late in replying during this coming week. Apologies for that.

**Comment by jonaslagoni** [comment]
> I have some doubts regarding [`spec-json-schema`](https://github.com/asyncapi/spec-json-schemas/blob/master/common/avroSchema_v1.json)
> 
>     * are we gonna implement `spec-json-schema` validation instead of [Avro Schema](https://github.com/asyncapi/modelina/blob/73f47d7f7a326cb7f40e71658574c3b9e9510f71/src/models/AvroSchema.ts) defined here or we just need to implement the scenario/concept used there for `avro array` and use that implementation while getting the Array Meta Model
> 
>     * not understood this particular thing, what you are referring to in here
>       > I suggest we focus on that
> 
> 
> **Note**: I have some academic exams and I might be late in replying during this coming week. Apologies for that.

The only reason why I linked the spec-json-schema files, was to show the structure of how an array is defined. 

But I guess we could validate the input as well 🤔 I would forget about it in this PR though.

No worries, no rush :v: Good luck in the exams!

**Comment by akkshitgupta** [comment]
Hey @jonaslagoni, I am getting an error whenever I run the integration test using [test/processors/AvroSchemaInputProcessor.spec.ts](https://github.com/asyncapi/modelina/pull/1753/files#diff-693fcd9d000d70a73213d420139079139158fcfeec401cf6fe58a1ae3cbe82cd). 

image

However, the [unit tests](https://github.com/asyncapi/modelina/pull/1753/files#diff-dccf6c2fcd45a4085d3b5269e63204dabee1600370f3ff23af0161f9c7f8d32e) are running fine and there is no typeerror. I tried to google the solution but could find success. can you help me with this particular thing 😄

**Comment by jonaslagoni** [comment]
@akkshitgupta what happens when you try running `includes` on the following type? https://github.com/asyncapi/modelina/pull/1753/files#diff-92b54f6485623708e919c7944aa95ef0c21d22401eeb4d2ee3c35cdafd57eb1eR24

File: src/processors/AvroSchemaInputProcessor.ts
Comment (jonaslagoni) [inline src/processors/AvroSchemaInputProcessor.ts:30]:
```suggestion
```

File: src/processors/InputProcessor.ts
Comment (jonaslagoni) [inline src/processors/InputProcessor.ts:23]:
```suggestion
    this.setProcessor('avro', new AvroSchemaInputProcessor());
```

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
I dont see any problems with your implementation, awesome @akkshitgupta  💪 

Now we just need to setup everything for the user to know about this new feature!

- Add [an example](https://github.com/asyncapi/modelina/blob/master/docs/contributing.md#adding-examples) such as for OpenAPI: https://github.com/asyncapi/modelina/tree/master/examples/openapi-from-object
- Add the example to the list of examples: https://github.com/asyncapi/modelina/blob/master/examples/README.md#input-examples
- Add it a section in the usage documentation: https://github.com/asyncapi/modelina/blob/master/docs/usage.md
- Add Avro to the list of supported inputs: https://github.com/asyncapi/modelina/blob/master/README.md#features

That should be it 💪

File: examples/avro-schema-from-object/index.ts
Comment (jonaslagoni) [inline examples/avro-schema-from-object/index.ts:71]: I would use a simple example here so it only outputs a single model, because the website dont expose it and we dont test it 🙂 

```suggestion
const AvroSchemaDoc = {
  name: 'Person',
  namespace: 'com.company',
  type: 'record',
  fields: [
    { name: 'name', type: 'string', example: 'Donkey', minLength: 0 },
    { name: 'serialNo', type: 'string', minLength: 0, maxLength: 50 },
    {
      name: 'email',
      type: ['null', 'string'],
      example: 'donkey@asyncapi.com',
      pattern: '^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$'
    },
    {
      name: 'age',
      type: ['null', 'int'],
      default: null,
      example: 123,
      exclusiveMinimum: 0,
      exclusiveMaximum: 200
    },
    {
      name: 'certifications',
      type: {
        type: 'array',
        items: 'string',
        minItems: 1,
        maxItems: 500,
        uniqueItems: true
      }
    },
    {
      name: 'weight',
      type: 'float',
      example: 65.1,
      minimum: 0,
      maximum: 500
    },
    {
      name: 'height',
      type: 'double',
      example: 1.85,
      minimum: 0,
      maximum: 3.0
    },
    { name: 'someid', type: 'string', logicalType: 'uuid' }
  ]
};
```

File: examples/avro-schema-from-object/index.spec.ts
Comment (jonaslagoni) [inline examples/avro-schema-from-object/index.spec.ts:12]:
```suggestion
    expect(spy.mock.calls.length).toEqual(1);
```

File: src/models/AvroSchema.ts
Comment (jonaslagoni) [inline src/models/AvroSchema.ts:75]: Are these functions used anywhere? 🤔

File: README.md
Comment (jonaslagoni) [inline README.md:200]: No need to be that explicit for the version, as we technically support all versions, cause we dont check it anywhere 🤔 
```suggestion
    <td>We support the following Avro versions: <a href="./examples/avro-schema-from-object">v1.x</a></td>
```

File: docs/usage.md
Comment (jonaslagoni) [inline docs/usage.md:119]: Incorrect, we support enum as well, is there any other types that you can define with Avro that we dont support after merging this?

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Got a few comments :v:

**Comment by jonaslagoni** [comment]
@akkshitgupta remember to resolve review comments when you have addressed them one way or another, and just ping me when you need another review :v:

**Comment by akkshitgupta** [comment]
@jonaslagoni can you have a look at the confusion I mentioned above 🙏🏻

**Comment by jonaslagoni** [comment]
@akkshitgupta I cannot see your comment

**Comment by akkshitgupta** [comment]
@jonaslagoni these two

https://github.com/asyncapi/modelina/pull/1753#discussion_r1564241626
https://github.com/asyncapi/modelina/pull/1753#discussion_r1564548748

**Comment by jonaslagoni** [comment]
@akkshitgupta for some reason I cannot see them at all 🤨 

Do you mind posting them again or as a comment? 🤔

**Comment by akkshitgupta** [comment]
Here are the comment images @jonaslagoni 😅 

image
image

**Comment by jonaslagoni** [comment]
@akkshitgupta do you see the `pending` label?  It means that you wrote those comments as part of a review that you have not published. You have to add them as single comments 🙂

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [reply src/helpers/AvroToMetaModel.ts]: This one is a bit hard, because we have two inputs that have a `type` keyword. JSON Schema and Avro, each with overlapping types such as `string` and `boolean`.

So the answer to the question lies within "what is the difference between Avro and JSON Schema and how can you easily detect it" 😄 

I don't have the perfect answer here unfortunately, other then look at the keywords that are different between them and match with those 🙂

File: src/models/AvroSchema.ts
Comment (jonaslagoni) [reply src/models/AvroSchema.ts:75]: Yes, please remove all code that is not in use or have an intended use 🙂

File: docs/usage.md
Comment (jonaslagoni) [reply docs/usage.md:119]: My question is, why are you making this `Note`, and which cases of Avro are we NOT supporting since you highlight `record` here? 🤨

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [reply src/helpers/AvroToMetaModel.ts]: > @jonaslagoni How about introducing a check explicitly in the shouldProcess() function using an additional schema property with a value of Avro to differentiate an Avro Schema before processing the input?

Can you give an example of what you mean?

File: docs/usage.md
Comment (akkshitgupta) [reply docs/usage.md:119]: In Avro Schema, a JSON Object is defined using one of the two datatypes: [`map`](https://avro.apache.org/docs/1.11.1/specification/#maps) and [`record`](https://avro.apache.org/docs/1.11.1/specification/#schema-record)

We do not handle the `map` datatype as I am not able to think of any use case. What would we been seeking in input object with single datatype. This is the reason I mentioned this explicitly in note.
image

File: src/helpers/AvroToMetaModel.ts
Comment (akkshitgupta) [reply src/helpers/AvroToMetaModel.ts]: Example: 

{
  "name": "Person",
  "type": "int",
  "schema": "avro"
}

Here, we can add a check for the `schema` property against the value `avro`. This would be to verify that only avro schema is passed to the Avro processor.

File: docs/usage.md
Comment (jonaslagoni) [reply docs/usage.md:119]: Maps can easily be converted to our DictionaryMetaModel, with key being StringMetaModel and the value what ever is defined 🙂

File: src/helpers/AvroToMetaModel.ts
Comment (jonaslagoni) [reply src/helpers/AvroToMetaModel.ts]: That would never happen no, because `schema` is not part of the Avro standard and never will be probably 🙂 

It's okay if the solution is not bullet proof, as long as we document the edge cases.

File: docs/usage.md
Comment (akkshitgupta) [reply docs/usage.md:119]: Didn't know this could be converted to `DictionaryMetaModel`.

```json
{
   "name": "additionalInfo",
   "type": "map",
   "value": "string"
}
```

Just wanna check if there is any use case of `fixed` and `byte` datatype defined in Avro Schema or not.

Comment (jonaslagoni) [reply docs/usage.md:119]: Why is the Avro specification so bad at explaining what is possible... Can you have the `value` be an array or record? 🤨 Or is only [primitive types](https://avro.apache.org/docs/1.11.1/specification/#primitive-types) allowed? 

`fixed` and `byte` I would convert to string models. We don't have official byte representation in the core models.

Comment (akkshitgupta) [reply docs/usage.md:119]: I need to check for this non primitive value type. However in my opinion I don't think non primitive is supposed to be there else it would have been explicitly mentioned there as done for the fields attribute in record type 

https://avro.apache.org/docs/1.11.1/specification/#schema-record

**Comment by jonaslagoni** [comment]
@akkshitgupta is it something you want to include in this PR, or want to merge this one and add issues to describe what is missing?

**Comment by akkshitgupta** [comment]
hello @jonaslagoni I have not been active for the past few days for some reason and also have my end-of-semester exams in the coming week. It would be great to merge this PR with other PRs to mention related issues, if that makes sense 😅😅. However, I would make sure to complete this ASAP. Thanks a lot for your kind support in my journey 😄

**Comment by jonaslagoni** [comment]
@akkshitgupta do you mind listing all the current short comings of the implementation right now?

**Comment by akkshitgupta** [comment]
Here are the pending tasks 

- implementation for `map`, `fixed` and `byte` data type
- find a work around to differentiate `Avro Schema` from `JSON Schema`
- update `docs/usage.md` with the edge cases and other related things.

**Comment by jonaslagoni** [comment]
>Here are the pending tasks
> - implementation for map, fixed and byte data type
> - find a work around to differentiate Avro Schema from JSON Schema
> - update docs/usage.md with the edge cases and other related things.


@akkshitgupta I am gonna merge this into the `next` branch, do you have time to add issues for each of those tasks, or are you full with stuff to do in school? :v:

**Comment by jonaslagoni** [comment]
Btw, you are free to join as a code owner for the Avro input processor if you want to 💪

**Review by jonaslagoni — APPROVED** [review]
Love the perseverance @akkshitgupta, thanks for sticking with the PR and fighting through the difficulties 💪

**Comment by jonaslagoni** [comment]
@all-contributors please add @akkshitgupta for code, example, test, docs

**Comment by akkshitgupta** [comment]
> Btw, you are free to join as a code owner for the Avro input processor if you want to 💪 

Thanks a lot @jonaslagoni for the opportunity. I would love to become a code owner. 🙏🏻

**Comment by jonaslagoni** [comment]
> Thanks a lot @jonaslagoni for the opportunity. I would love to become a code owner. 🙏🏻

Go ahead and add the Avro input processor to the codeowners file and add yourself ✌️ https://github.com/asyncapi/modelina/blob/e4a3e4e66acb70f9b00b0e2bb45701280c36aaa9/CODEOWNERS#L19

**Pull request merged**
